Assets/MRTK/SDK/Features/UX/Scripts/InteractiveElement/States/StateManager.cs

```diff
@@ -0,0 +1,422 @@
+        {
+            states = trackedStates.States;
+
+            interactiveElement = interactiveElementSource;
+
+            InitializeEventReceiverManager();
+        }
+
+        // List of states to be tracked by this state manager
+        private List<InteractionState> states = null;
+
+        /// <summary>
+        /// The read only list of the current Interaction States being tracked.  To modify this list use the AddNewState()
+        /// RemoveState() methods, to set the value of a state in this list use SetStateOn/Off() methods.
+        /// </summary>
+        public IList<InteractionState> States => states.AsReadOnly();
+
+        /// <summary>
+        /// The Event Receiver Manager for this State Manager. Each state can contain an event configuration scriptable which defines
+        /// the events associated with the state.  The Event Receiver Manager depends on a State Manager.
+        /// </summary>
+        public EventReceiverManager EventReceiverManager { get; internal set; } = null;
+
+        // The interactive element for this state manager
+        private BaseInteractiveElement interactiveElement = null;
+
+        /// <summary>
+        /// The Unity Event with the activated state as the event data. This event is invoked when a state is 
+        /// set to on.
+        /// </summary>
+        public InteractionStateActiveEvent OnStateActivated { get; protected set; } = new InteractionStateActiveEvent();
+
+        /// <summary>
+        /// The Unity Event with the previous active state and the current active state.  The event is invoked when 
+        /// a state is set to off.
+        /// </summary>
+        public InteractionStateInactiveEvent OnStateDeactivated { get; protected set; } = new InteractionStateInactiveEvent();
+
+        // List of all core states
+        private string[] coreStates = Enum.GetNames(typeof(CoreInteractionState)).ToArray();
+
+        // List of active states, used for tracking the current and previous states
+        private List<InteractionState> activeStates = new List<InteractionState>();
+
+        // Initializes the EventReceiverManager and creates the runtime classes for states that contain a valid 
+        // configuration
+        internal void InitializeEventReceiverManager()
+        {
+            // Create a new event receiver manager for this state manager
+            EventReceiverManager = new EventReceiverManager(this);
+            EventReceiverManager.InitializeEventReceivers();
+        }
+
+        /// <summary>
+        /// Gets a state by using the state name.
+        /// </summary>
+        /// <param name="stateName">The name of the state to retrieve</param>
+        /// <returns>The state contained in the Tracked States scriptable object.</returns>
+        public InteractionState GetState(string stateName)
+        {
+            InteractionState interactionState = states.Find((state) => state.Name == stateName);
```

Comment:
If this runs every frame (I think it could), consider using a different data type, like a dictionary<string, InteractionState>. At least make sure that the states data structure uses binary search (by default I think find is linear since the list is not guaranteed to be sorted). Seems like dictionary would be the way to go though...

Reply:
A dictionary is much better in this case. Added.